### DISCUSSION TITLE
UX improvement feature suggestions

### DISCUSSION THREAD

**Discussion by Brookke** [aviator-co/av#302] [Ideas]
Firstly thank you so much for making this tool. I think it's great. I've been using it for a couple of days and have these suggestions that I think could really help with the workflow. Was unsure if this was the best place to post it, feel free to let me know if not and I will move it.

## Allow `av stack sync` to be ran at the root of the repo
Reason: So you can sync all your stacks in one go rather than having to do them individually. Additionally you would be able to easily clean up all merged branches with `--prune` in one go

## Make `av stack sync` run --trunk, --prune, by default in an interactive way
For example:
1. run `av stack sync`
2. prompt `Branch X has been merged do you wish to delete this branch y/n`
3. prompt `This stack is behind your main branch do you wish to sync to the trunk y/n`

Reason: it would keep the repo clean with less conscious effort to run prune and trunk commands 

## Make `av stack submit` check if the stack needs syncing before pushing
For example:
1. run `av stack submit`
2. prompt `Your branch X has fallen behind its parent, do you want to restack before submitting y/n`

Reason: stops out of data PRs from being created

## Improvements to navigating around stacks (Done)
- ~Make `av stack prev` move to trunk when at the bottom of the stack rather than erroring~ (done #317)
- ~Add `av stack top` and `av stack bottom` for moving to either end of the stack quickly~
    - these are covered already by `av stack next --last` so this isn't required
- ~Make `av stack next` let you select when there are multiple options~ (done #303)

## Add top level aliases for common commands
For example:
- `av next`  -> `av stack next`
- `av prev` -> `av stack prev`
- `av submit` -> `av stack submit`
- `av modify` -> `av commit amend --no-edit`
- `av amend` -> `av commit amend`
- `ac create` -> `av stack branch`

Reason: better out of box experience, faster to preform these common actions - I realise I could make these with git aliases (and I have), but would be great if they were out of the box

##  ~Automatically outline the PR stack in the PR description~ (Done)

~For example:~

- ~👉🏻 #1234 feat: adding the component~
- ~#1235 feat: wiring it up to the backend~
- ~#1236 feat: adding to correct pages~

**Comment by draftcode** [comment]
> Allow av stack sync to be ran at the root of the repo

Probably what you are looking for is `--all`.

> Make av stack sync run --trunk, --prune, by default in an interactive way

This is already planned, but in a slightly different way. We are planning to split `av stack sync` into different commands as it does 3-4 different things in one command. Our planned split is:

* `av stack restack` to have `--trunk=false --no-fetch --no-push` like behavior.
* `av stack reparent` to have `--parent --no-fetch --no-push` like behavior.
* `av stack adopt` to have the same effect as reparent, but smart stack adoption (== checking the commit structure to figure out who is the parent and who are the children).

Then, after this, we will make `av stack sync` to have `--trunk` default and drop `--no-fetch` (If people want `--no-fetch`, use restack). We are aiming that `sync` command to `sync` with GitHub. In addition to this, we want to change the prune behavior to ask (which is the same as you said). This will be configurable (== the config can make it yes/no/ask).

We've already done `restack` and `adopt` part. `reparent` is coming, and then we can finally work on `sync` v2. In v2, we also make it batch fetch/push. git-push can push multiple refs at the same time, so we can push them altogether, making sync ops much faster.

> Make av stack prev move to trunk when at the bottom of the stack rather than erroring

Sure.

> Add top level aliases for common commands

Let's hold on this one. Yes. Most of the commands are anyway under `av stack`, but want to be careful about aliasing / renaming to avoid confusion / conflict.

As you found out, aliasing as a git subcommand is something we suggest (https://docs.aviator.co/aviator-cli/how-to-guides/git-subcommand-aliasing). I, for example, set `git sync` to `av stack sync --prune --trunk --all`.

> Automatically outline the PR stack in the PR description

This is done at https://github.com/aviator-co/av/pull/271.

> Add av stack absob similar to [git-absorb](https://github.com/tummychow/git-absorb) or [sapling absorb](https://sapling-scm.com/docs/commands/absorb/)

This is also planned.
